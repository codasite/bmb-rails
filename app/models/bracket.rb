class Bracket < ApplicationRecord
  validates :title, presence: true
  validates :slug, presence: true, uniqueness: true
  validates :num_teams, presence: true, numericality: { greater_than: 0 }

  before_validation :generate_slug, if: :title_changed?

  serialize :round_names, coder: JSON

  scope :by_year, ->(year) { where(year: year) if year.present? }
  scope :by_month, ->(month) { where(month: month) if month.present? }
  scope :voting, -> { where(is_voting: true) }
  scope :tournament, -> { where(is_voting: false) }

  has_many :bracket_matches, dependent: :destroy
  has_many :picks, dependent: :destroy
  has_many :plays, dependent: :destroy

  def to_param
    slug
  end

  def voting?
    is_voting
  end

  def tournament?
    !is_voting
  end

  def has_fee?
    fee.present? && fee > 0
  end

  def current_round
    live_round_index
  end

  def completed?
    # Logic to determine if bracket is completed
    # This would depend on your specific business rules
    false
  end

  private

  def generate_slug
    self.slug = title.parameterize if title.present?
  end
end
