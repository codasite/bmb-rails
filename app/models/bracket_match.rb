class BracketMatch < ApplicationRecord
  belongs_to :bracket
  belongs_to :team1, class_name: 'Team', optional: true
  belongs_to :team2, class_name: 'Team', optional: true
  belongs_to :winner, class_name: 'Team', optional: true

  has_many :picks, dependent: :destroy

  validates :round, presence: true
  validates :position, presence: true

  scope :by_round, ->(round) { where(round: round) }
  scope :completed, -> { where.not(winner: nil) }
  scope :pending, -> { where(winner: nil) }

  def completed?
    winner.present?
  end

  def pending?
    !completed?
  end

  def teams
    [team1, team2].compact
  end

  def has_both_teams?
    team1.present? && team2.present?
  end
end
