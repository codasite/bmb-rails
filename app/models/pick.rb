class Pick < ApplicationRecord
  belongs_to :bracket
  belongs_to :user, optional: true
  belongs_to :team
  belongs_to :bracket_match

  validates :round, presence: true
  validates :position, presence: true

  scope :by_round, ->(round) { where(round: round) }
  scope :by_user, ->(user) { where(user: user) }
  scope :results, -> { where(is_result: true) }
  scope :predictions, -> { where(is_result: false) }

  def result?
    is_result
  end

  def prediction?
    !is_result
  end
end
