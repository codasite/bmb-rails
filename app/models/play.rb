class Play < ApplicationRecord
  belongs_to :bracket
  belongs_to :user

  validates :score, numericality: { greater_than_or_equal_to: 0 }, allow_nil: true

  serialize :picks_data, coder: JSON

  scope :by_user, ->(user) { where(user: user) }
  scope :by_bracket, ->(bracket) { where(bracket: bracket) }
  scope :completed, -> { where.not(completed_at: nil) }
  scope :active, -> { where(completed_at: nil) }

  def completed?
    completed_at.present?
  end

  def active?
    !completed?
  end

  def calculate_score
    # Logic to calculate score based on picks vs results
    # This would be implemented based on your scoring rules
    0
  end
end
