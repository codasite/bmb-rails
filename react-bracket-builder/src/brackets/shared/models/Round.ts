import { Nullable } from '../../../utils/types'
import { MatchNode } from './operations/MatchNode'
import { MatchRepr } from '../api/types/bracket'
import { Team } from './Team'

export class Round {
  index: number
  depth: number
  matches: Array<Nullable<MatchNode>>
  constructor(
    index: number,
    depth: number,
    matches: Array<Nullable<MatchNode>> = []
  ) {
    this.index = index
    this.depth = depth
    this.matches = matches
  }
  allPicked(): boolean {
    return this.matches.every((match) => {
      if (match === null) {
        return true
      }
      return match.team1Wins || match.team2Wins
    })
  }
  getTeam(teamId: number): Nullable<Team> {
    for (const [index, match] of this.matches.entries()) {
      if (match === null) {
        continue
      }
      const team1 = match.getTeam1()
      const team2 = match.getTeam2()

      if (team1?.id === teamId) {
        team1.side = index < this.matches.length / 2 ? 'left' : 'right'
        return team1
      }

      if (team2?.id === teamId) {
        team2.side = index < this.matches.length / 2 ? 'left' : 'right'
        return team2
      }
    }

    return null
  }
  serialize(): Nullable<MatchRepr>[] {
    return this.matches.map((match, i) => {
      if (match === null) {
        return null
      }
      return match.serialize()
    })
  }
}
