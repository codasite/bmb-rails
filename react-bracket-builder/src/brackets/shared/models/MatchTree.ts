import { Nullable } from '../../../utils/types'
import { BracketRes, MatchPick, MatchReq, MatchTreeRepr } from '../api'
import { WildcardPlacement } from './WildcardPlacement'
import { Team } from './Team'
import { MatchNode } from './operations/MatchNode'
import { Round } from './Round'
import { matchReprFromNumTeams } from './operations/MatchReprFromNumTeams'
import { getNumRounds } from './operations/GetNumRounds'
import { matchReprFromRes } from './operations/MatchReprFromRes'
import { linkNodes } from './operations/LinkNodes'

export class MatchTree {
  rounds: Round[]
  private numTeams: number
  private wildcardPlacement?: WildcardPlacement
  isVoting: boolean
  liveRoundIndex: number

  constructor(
    rounds: Round[] = [],
    wildcardPlacement?: WildcardPlacement,
    isVoting: boolean = false,
    liveRoundIndex: number = 0
  ) {
    linkNodes(rounds)
    this.rounds = rounds
    this.wildcardPlacement = wildcardPlacement
    this.isVoting = isVoting
    this.liveRoundIndex = liveRoundIndex
  }

  getRootMatch(): Nullable<MatchNode> {
    const lastRound = this.rounds[this.rounds.length - 1]
    if (!lastRound) {
      return null
    }
    return lastRound.matches[0]
  }

  clone(): MatchTree {
    return MatchTree.deserialize(this.serialize())
  }

  equals(other: MatchTree): boolean {
    return (
      JSON.stringify(this.serialize()) === JSON.stringify(other.serialize())
    )
  }

  /**
   * Returns the total number of POSSIBLE teams in the tournament.
   * This should reflect the number of teams that would be used to create the same structure with MatchTree.fromNumTeams()
   */
  getNumTeams(): number {
    if (this.numTeams) {
      return this.numTeams
    }
    // count all non null matches in the tree
    const numMatches = this.rounds.reduce((acc, round) => {
      return (
        acc +
        round.matches.reduce((acc, match) => {
          if (match) {
            return acc + 1
          }
          return acc
        }, 0)
      )
    }, 0)

    const numTeams = numMatches + 1
    this.numTeams = numTeams

    return numTeams
  }

  getWildcardPlacement(): WildcardPlacement | undefined {
    return this.wildcardPlacement
  }

  serialize(): MatchTreeRepr {
    const tree = this
    const rounds = tree.rounds.map((round) => {
      return round.serialize()
    })
    return {
      rounds: rounds,
      wildcardPlacement: tree.wildcardPlacement,
      isVoting: tree.isVoting,
      liveRoundIndex: tree.liveRoundIndex,
    }
  }

  advanceTeam = (
    roundIndex: number,
    matchIndex: number,
    left: boolean,
    requireComplete: boolean = false
  ) => {
    if (requireComplete && roundIndex > 0) {
      const prevRound = this.rounds[roundIndex - 1]
      if (prevRound && !prevRound.allPicked()) {
        return
      }
    }
    const round = this.rounds[roundIndex]
    const match = round.matches[matchIndex]
    if (!match) {
      return
    }
    match.team1Wins = false
    match.team2Wins = false
    if (left) {
      match.team1Wins = true
    } else {
      match.team2Wins = true
    }
  }

  anyPicked = (): boolean => {
    return this.rounds.some((round) => {
      return round.matches.some((match) => {
        if (!match) {
          return false
        }
        return match.team1Wins || match.team2Wins
      })
    })
  }

  allPicked = (): boolean => {
    if (this.isVoting) {
      return this.rounds[this.liveRoundIndex].allPicked()
    }
    return this.rounds.every((round) => {
      return round.allPicked()
    })
  }

  allTeamsAdded = (): boolean => {
    return this.everyMatch((match) => {
      let hasNeededTeams = true
      if (!match.left && !match.getTeam1()?.name) {
        hasNeededTeams = false
      }
      if (hasNeededTeams && !match.right && !match.getTeam2()?.name) {
        hasNeededTeams = false
      }
      return hasNeededTeams
    })
  }

  toMatchReq = (): MatchReq[] => {
    const matches: MatchReq[] = []
    this.forEachMatch((match, matchIndex, roundIndex) => {
      const matchReq: MatchReq = {
        roundIndex: match.roundIndex,
        matchIndex: match.matchIndex,
      }
      if (!match.left) {
        matchReq.team1 = match.getTeam1()?.toTeamReq()
      }
      if (!match.right) {
        matchReq.team2 = match.getTeam2()?.toTeamReq()
      }
      if (matchReq.team1 || matchReq.team2) {
        matches.push(matchReq)
      }
    })
    return matches
  }

  toMatchPicks = (): MatchPick[] => {
    const picks: MatchPick[] = []
    this.rounds.forEach((round, roundIndex) => {
      round.matches.forEach((match, matchIndex) => {
        if (!match) {
          return
        }
        const { team1Wins, team2Wins } = match
        const team1 = match.getTeam1()
        const team2 = match.getTeam2()
        const team1Winner = team1Wins && team1
        const team2Winner = team2Wins && team2

        if (team1Winner && team1.id) {
          picks.push({
            roundIndex,
            matchIndex,
            winningTeamId: team1.id,
          })
        } else if (team2Winner && team2.id) {
          picks.push({
            roundIndex,
            matchIndex,
            winningTeamId: team2.id,
          })
        }
      })
    })
    return picks
  }

  forEachMatch = (
    callback: (match: MatchNode, matchIndex: number, roundIndex: number) => void
  ) => {
    this.rounds.forEach((round, roundIndex) => {
      round.matches.forEach((match, matchIndex) => {
        if (!match) {
          return
        }
        callback(match, matchIndex, roundIndex)
      })
    })
  }

  everyMatch = (
    callback: (
      match: MatchNode,
      matchIndex: number,
      roundIndex: number
    ) => boolean
  ) => {
    return this.rounds.every((round, roundIndex) => {
      return round.matches.every((match, matchIndex) => {
        if (!match) {
          return true
        }
        return callback(match, matchIndex, roundIndex)
      })
    })
  }

  findMatch = (
    callback: (match: Nullable<MatchNode>) => boolean
  ): Nullable<MatchNode> => {
    let foundMatch: Nullable<MatchNode> = null
    this.rounds.some((round) => {
      return round.matches.some((match) => {
        if (callback(match)) {
          foundMatch = match
          return true
        }
        return false
      })
    })
    return foundMatch
  }

  static fromNumTeams(
    numTeams: number,
    wildcardPlacement: WildcardPlacement = WildcardPlacement.Top
  ): MatchTree {
    const matches = matchReprFromNumTeams(numTeams, wildcardPlacement)
    return MatchTree.deserialize({
      rounds: matches,
      wildcardPlacement,
    })
  }

  static fromMatchRes(
    bracket: Pick<
      BracketRes,
      'matches' | 'numTeams' | 'isVoting' | 'liveRoundIndex'
    >,
    wildcardPlacement?: WildcardPlacement
  ): MatchTree | null {
    const numRounds = getNumRounds(bracket.numTeams)

    const nestedMatches = matchReprFromRes(numRounds, bracket.matches)
    return MatchTree.deserialize({
      rounds: nestedMatches,
      wildcardPlacement,
      isVoting: bracket.isVoting,
      liveRoundIndex: bracket.liveRoundIndex,
    })
  }

  static fromPicks(
    bracket: Pick<
      BracketRes,
      'matches' | 'numTeams' | 'isVoting' | 'liveRoundIndex'
    >,
    picks: MatchPick[],
    wildcardPlacement?: WildcardPlacement
  ): MatchTree | null {
    const matchTree = MatchTree.fromMatchRes(bracket, wildcardPlacement)
    if (!matchTree) {
      return null
    }
    for (const pick of picks) {
      const { roundIndex, matchIndex, winningTeamId } = pick
      const match = matchTree.rounds[roundIndex].matches[matchIndex]
      if (!match) {
        throw new Error('Match not found')
      }
      match._pick = pick

      const team1 = match.getTeam1()
      const team2 = match.getTeam2()
      if (team1?.id === winningTeamId) {
        match.team1Wins = true
      } else if (team2?.id === winningTeamId) {
        match.team2Wins = true
      } else {
        throw new Error('Match does not have winningTeam from pick')
      }
    }
    return matchTree
  }

  static deserialize(matchTreeRepr: MatchTreeRepr): MatchTree {
    const { rounds: matchRes, wildcardPlacement } = matchTreeRepr
    const rounds = matchRes.map((round, roundIndex) => {
      const depth = matchRes.length - roundIndex - 1
      const newRound = new Round(roundIndex, depth)
      const matches = round.map((match, matchIndex) => {
        if (match === null) {
          return null
        }
        const {
          id,
          team1: team1Repr,
          team2: team2Repr,
          team1Wins,
          team2Wins,
          pick,
        } = match
        const team1 = team1Repr ? new Team(team1Repr.name, team1Repr.id) : null
        const team2 = team2Repr ? new Team(team2Repr.name, team2Repr.id) : null
        // const newMatch = new MatchNode(roundIndex, matchIndex, depth, match.id, team1, team2, team1Wins, team2Wins)
        const newMatch = new MatchNode({
          id,
          matchIndex,
          roundIndex,
          team1,
          team2,
          team1Wins,
          team2Wins,
          left: null,
          right: null,
          parent: null,
          depth,
          pick,
        })
        return newMatch
      })
      newRound.matches = matches
      return newRound
    })
    const tree = new MatchTree(
      rounds,
      wildcardPlacement,
      matchTreeRepr.isVoting,
      matchTreeRepr.liveRoundIndex
    )
    return tree
  }
}
