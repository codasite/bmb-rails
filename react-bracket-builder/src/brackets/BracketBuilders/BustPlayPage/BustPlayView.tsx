import { useEffect, useState, useContext } from 'react'
import * as Sentry from '@sentry/react'
import { MatchTree } from '../../shared/models/MatchTree'
import { BusterBracket } from '../../shared/components/Bracket'
import { ActionButton } from '../../shared/components/ActionButtons'
import redBracketBg from '../../shared/assets/bracket-bg-red.png'
import { bracketApi } from '../../shared/api/bracketApi'
import { BracketRes, PlayReq, PlayRes } from '../../shared/api/types/bracket'
import { BusterVsBustee } from './BusterVersusBustee'
import { useWindowDimensions } from '../../../utils/hooks'
import { getBracketWidth } from '../../shared/components/Bracket/utils'
import { getNumRounds } from '../../shared/models/operations/GetNumRounds'
import { PaginatedBuilder } from '../PaginatedBuilderBase/PaginatedBuilder'
import { getBustTrees } from './utils'
import { BustEndPage, BustBracketPages } from './PaginatedBustBuilder'
import { BracketBusterContext } from './context'
import { WindowDimensionsContext } from '../../shared/context/WindowDimensionsContext'

interface BustPlayViewProps {
  bracket?: BracketRes
  busterTree?: MatchTree
  setBusterTree?: (matchTree: MatchTree) => void
  busteeDisplayName: string
  busteeThumbnail: string
  onButtonClick?: () => void
  buttonText?: string
  processing?: boolean
}

export const BustPlayView = (props: BustPlayViewProps) => {
  const { width: windowWidth, height: windowHeight } = useContext(
    WindowDimensionsContext
  )
  const {
    bracket,
    busterTree,
    setBusterTree,
    busteeDisplayName,
    busteeThumbnail,
    onButtonClick,
    buttonText,
    processing,
  } = props

  const showPaginated =
    windowWidth - 100 < getBracketWidth(getNumRounds(bracket?.numTeams))

  if (showPaginated && busterTree) {
    return (
      <BracketBusterContext.Provider
        value={{
          busteeDisplayName,
          busteeThumbnail,
          buttonText,
        }}
      >
        <PaginatedBuilder
          EndPageComponent={BustEndPage}
          BracketPagesComponent={BustBracketPages}
          handleSubmit={onButtonClick}
          matchTree={busterTree}
          setMatchTree={setBusterTree}
        />
      </BracketBusterContext.Provider>
    )
  }

  return (
    <div
      className={`wpbb-reset tw-uppercase tw-bg-no-repeat tw-bg-top tw-bg-cover`}
      style={{
        backgroundImage: `url(${redBracketBg})`,
      }}
    >
      <div
        className={`tw-flex tw-flex-col tw-items-center tw-max-w-screen-lg tw-mx-auto tw-gap-50 tw-pb-100 tw-pt-40`}
      >
        {busterTree && (
          <>
            <BusterVsBustee
              busteeDisplayName={busteeDisplayName}
              busteeThumbnail={busteeThumbnail}
            />
            <BusterBracket
              matchTree={busterTree}
              setMatchTree={setBusterTree}
            />
            <div className="tw-flex tw-flex-col tw-justify-center tw-items-center">
              <ActionButton
                variant="red"
                size="big"
                disabled={!busterTree?.allPicked() || processing}
                onClick={onButtonClick}
              >
                {buttonText}
              </ActionButton>
            </div>
          </>
        )}
      </div>
    </div>
  )
}
