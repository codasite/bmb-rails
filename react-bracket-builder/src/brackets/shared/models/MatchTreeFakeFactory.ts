import { MatchTree } from './MatchTree'

export function fourTeamMatchTree() {
  const matches = [
    [
      {
        id: 9,
        roundIndex: 0,
        matchIndex: 0,
        team1: { id: 17, name: 'Team 1' },
        team2: { id: 18, name: 'Team 2' },
        team1Wins: true,
      },
      {
        id: 10,
        roundIndex: 0,
        matchIndex: 1,
        team1: { id: 19, name: 'Team 3' },
        team2: { id: 20, name: 'Team 4' },
        team1Wins: true,
      },
    ],
    [{ roundIndex: 1, matchIndex: 1, team2Wins: true }],
  ]
  return MatchTree.deserialize({ rounds: matches })
}
