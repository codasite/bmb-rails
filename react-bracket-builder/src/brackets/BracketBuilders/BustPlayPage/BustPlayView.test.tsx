// DO NOT REMOVE REACT IMPORT. Needed for image generator
import React from 'react'
import { render } from '@testing-library/react'
import { BustPlayView } from './BustPlayView'
import { fourTeamMatchTree } from '../../shared/models/MatchTreeFakeFactory'
import {
  MatchTreeContext1,
  MatchTreeContext2,
} from '../../shared/context/context'

describe('BustPlayView', () => {
  test('renders page correctly', () => {
    const noop = () => {}
    const { asFragment } = render(
      <MatchTreeContext1.Provider
        value={{ matchTree: fourTeamMatchTree(), setMatchTree: noop }}
      >
        <MatchTreeContext2.Provider value={{ matchTree: fourTeamMatchTree() }}>
          <BustPlayView
            busterTree={fourTeamMatchTree()}
            busteeDisplayName={'test name'}
            busteeThumbnail={'test'}
            buttonText={'button text'}
          />
        </MatchTreeContext2.Provider>
      </MatchTreeContext1.Provider>
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
