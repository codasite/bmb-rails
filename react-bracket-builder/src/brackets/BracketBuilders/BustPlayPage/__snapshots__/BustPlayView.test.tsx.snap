// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BustPlayView renders page correctly 1`] = `
<DocumentFragment>
  <div
    class="wpbb-reset tw-uppercase tw-bg-no-repeat tw-bg-top tw-bg-cover"
    style="background-image: url(undefined);"
  >
    <div
      class="tw-flex tw-flex-col tw-items-center tw-max-w-screen-lg tw-mx-auto tw-gap-50 tw-pb-100 tw-pt-40"
    >
      <div
        class="tw-flex tw-gap-18 tw-justify-center tw-items-center"
      >
        <div
          class="tw-flex tw-flex-col tw-items-center tw-gap-4"
        >
          <img
            alt="celebrity-photo"
            class="tw-h-52 tw-w-52 tw-rounded-full tw-shadow-lg tw-shadow-blue"
            src="test"
          />
          <span
            class="tw-text-white tw-font-700 tw-text-12"
          >
            test name
          </span>
        </div>
        <span
          class="tw-text-white tw-font-700 tw-text-48"
        >
          VS
        </span>
        <div
          class="tw-flex tw-flex-col tw-items-center tw-gap-4"
        >
          <div
            class="tw-h-52 tw-w-52 tw-flex tw-rounded-full tw-border-solid tw-border-1 tw-border-red tw-justify-center tw-items-center tw-bg-red/15"
          >
            <svg
              class="tw-text-red"
            />
          </div>
          <span
            class="tw-text-white tw-font-700 tw-text-12"
          >
            You
          </span>
        </div>
      </div>
      <div
        class="tw-flex tw-flex-col tw-dark wpbb-default-bracket tw-relative"
      >
        <div
          class="tw-flex tw-flex-col tw-justify-end"
          style="margin-bottom: 95px; min-height: 0;"
        >
          <div
            class="tw-flex tw-flex-col tw-items-center"
            style="gap: 24px;"
          >
            <div
              class="wpbb-team-1-0-winner tw-relative tw-flex tw-justify-center tw-items-center tw-whitespace-nowrap tw-leading-none tw-uppercase tw-font-sans tw-w-[297px] tw-h-[52px] tw-text-white tw-font-700 tw-bg-red"
              style="border-width: 0px;"
            >
              <span
                style="font-size: 32px; transform: scale(1); visibility: hidden;"
              >
                Team 3
              </span>
            </div>
          </div>
        </div>
        <div
          class="tw-flex tw-flex-col tw-justify-center tw-h-100"
        >
          <div
            class="tw-flex tw-justify-between"
            style="width: 700px;"
          >
            <div
              class="wpbb-default-match-column tw-flex tw-flex-col tw-justify-center tw-gap-[80px]"
            >
              <div
                class="tw-flex tw-flex-col tw-gap-[20px] tw-translate-y-[0px]"
              >
                <div
                  class="wpbb-team-0-0-left tw-relative tw-flex tw-justify-center tw-items-center tw-whitespace-nowrap tw-leading-none tw-uppercase tw-font-sans tw-w-[115px] tw-h-[28px] tw-text-white tw-font-500 tw-bg-red"
                  style="border-width: 0px;"
                >
                  <span
                    style="font-size: 16px; transform: scale(1); visibility: hidden;"
                  >
                    Team 1
                  </span>
                </div>
                <div
                  class="wpbb-team-0-0-right tw-relative tw-flex tw-justify-center tw-items-center tw-whitespace-nowrap tw-leading-none tw-uppercase tw-font-sans tw-w-[115px] tw-h-[28px] tw-text-white tw-font-500 tw-bg-transparent tw-border-solid tw-border-white/50"
                  style="border-width: 2px;"
                >
                  <span
                    style="font-size: 16px; transform: scale(1); visibility: hidden;"
                  >
                    Team 2
                  </span>
                </div>
              </div>
            </div>
            <div
              class="wpbb-default-match-column tw-flex tw-flex-col tw-justify-center tw-gap-[236px]"
            >
              <div
                class="tw-flex tw-flex-col tw-gap-[20px] tw-translate-y-[-48px]"
              >
                <div
                  class="wpbb-team-1-0-left tw-relative tw-flex tw-justify-center tw-items-center tw-whitespace-nowrap tw-leading-none tw-uppercase tw-font-sans tw-w-[115px] tw-h-[28px] tw-text-white tw-font-500 tw-bg-transparent tw-border-solid tw-border-white/50"
                  style="border-width: 2px;"
                >
                  <span
                    style="font-size: 16px; transform: scale(1); visibility: hidden;"
                  >
                    Team 1
                  </span>
                </div>
                <div
                  class="wpbb-team-1-0-right tw-relative tw-flex tw-justify-center tw-items-center tw-whitespace-nowrap tw-leading-none tw-uppercase tw-font-sans tw-w-[115px] tw-h-[28px] tw-text-white tw-font-500 tw-bg-red"
                  style="border-width: 0px;"
                >
                  <span
                    style="font-size: 16px; transform: scale(1); visibility: hidden;"
                  >
                    Team 3
                  </span>
                </div>
              </div>
            </div>
            <div
              class="wpbb-default-match-column tw-flex tw-flex-col tw-justify-center tw-gap-[80px]"
            >
              <div
                class="tw-flex tw-flex-col tw-gap-[20px] tw-translate-y-[0px]"
              >
                <div
                  class="wpbb-team-0-1-left tw-relative tw-flex tw-justify-center tw-items-center tw-whitespace-nowrap tw-leading-none tw-uppercase tw-font-sans tw-w-[115px] tw-h-[28px] tw-text-white tw-font-500 tw-bg-red"
                  style="border-width: 0px;"
                >
                  <span
                    style="font-size: 16px; transform: scale(1); visibility: hidden;"
                  >
                    Team 3
                  </span>
                </div>
                <div
                  class="wpbb-team-0-1-right tw-relative tw-flex tw-justify-center tw-items-center tw-whitespace-nowrap tw-leading-none tw-uppercase tw-font-sans tw-w-[115px] tw-h-[28px] tw-text-white tw-font-500 tw-bg-transparent tw-border-solid tw-border-white/50"
                  style="border-width: 2px;"
                >
                  <span
                    style="font-size: 16px; transform: scale(1); visibility: hidden;"
                  >
                    Team 4
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          style="margin-top: 20px; min-height: 0;"
        >
          <div
            class="tw-flex tw-flex-col tw-gap-20 tw-justify-between tw-items-center tw-font-700 tw-whitespace-nowrap "
          >
            <span
              class="tw-text-36 tw-text-dd-blue dark:tw-text-white"
            >
              Who You Got?
            </span>
            <svg
              class="tw-w-[124px] tw-text-black/25 dark:tw-text-white"
            />
          </div>
        </div>
        <div
          class="tw-absolute"
        />
        <div
          class="tw-absolute"
        />
      </div>
      <div
        class="tw-flex tw-flex-col tw-justify-center tw-items-center"
      >
        <button
          class="tw-flex tw-flex-row tw-items-center tw-justify-center tw-font-sans tw-uppercase tw-whitespace-nowrap tw-cursor-pointer disabled:tw-cursor-default tw-leading-none tw-cursor-pointer tw-bg-red/20 tw-text-white tw-gap-10 tw-text-36 tw-font-700 tw-rounded-8 tw-px-60 tw-py-16 tw-border-solid tw-border-4 tw-border-red"
        >
          button text
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
